docs/01-tooling/01-automatic-file-bundling.md

```diff
@@ -1,6 +1,188 @@
+  production if the dependencies are not properly specified
+* If possible, it should be easier to tap into npm libraries for our front-end
+  needs, and check for updates
+
 ### Solutions
+
+We considered our options, and after discussion among the engineers, we decided
+that:
+
+* We would use a node based file bundler ([webpack][]), in order to:
+  * automatically bundle our JS sources with a single source of truth (the
+    source code itself)
+  * use a standard module system (we started with common.js and afterwards
+    migrated to ES modules when finalized)
+  * validate the dependency graph and asset building before going to production
+    (webpack statically analyzes the asset tree and fails on build)
+* We would introduce a build step to compile sources into production worthy
+  assets
+* We would use ResourceLoader as the way to **serve** the JS bundle(s) and to
+  specify other dependencies (images, i18n messages, styles)
+
+You can read some more details in the architecture design record [4. Use
+webpack][use-webpack] that we wrote when we discussed this in the team.
+
+#### Why webpack and not `<my-favorite-tool>`?
+
+We looked at the ecosystem of bundlers at the time, and this is a summary of
+our evaluation:
+
+* Browserify: Great tool, shrinking community and support in favor of others
+* Rollup: Great tool, specialized in bundling code for producing libraries, not
+  applications
+* Webpack: Great tool, community, maintainers, financial support, mindshare and
+  ecosystem of tools
+
+We are not married to, or using any webpack specific features, so we can migrate
+from this tool in the future if it becomes a problem.
+
 ### Results
+
+* We introduced `webpack` ([config][webpack-config]) and [build scripts][]
+* The JS files are located in [src/][src] and use EcmaScript modules
+  ([example][es-modules-example])
+* The production assets get built into [/resources/dist][dist] to be served by
+  ResourceLoader
+* We added a CI step in the `npm-test` job that checks built assets have been
+  commited and up to date ([check-built-assets][check-built-assets])
+* Added a precommit hook that automatically runs the build step and adds the
+  built assets ([precommit][precommit])
+
+These are some of the benefits we have seen after the introduction of these
+changes:
+
+* ResourceModules configuration reduced for JS files, file order is automatic
+  now, source code is the single truth for dependency order
+  * [Before][scripts-order], and [after][scripts-after] (no manual ordering of
+    files)
+* Decreased cost and burden for splitting modules, reusing code, adding new code
+  since file order is automatically sorted out
+* Source maps in development that point to the original modules in `src/` thanks
+  to webpack
+* Anecdotally, faster ResourceLoader performance by offloading file parsing and
+  bundling to the node CLI tool watcher (`npm start`)
+* We bundle `redux` and `redux-thunk` from npm, based on the [pinned
```

Comment:
We should point to the security review of these two libraries here <[T107561](https://phabricator.wikimedia.org/T107561)>. I think there'll be some concern that we're using libraries distributed by NPM, given the discussion on [T107561](https://phabricator.wikimedia.org/T107561).

We should also stress that this is a very simple change to reverse.

Reply:
I know, my spider sense tingled too, but to be honest, we have taken many precaution steps, and it is a social problem including libraries a go-go, and that is not what we've done.

We have the sha in package-lock.json, the security review, the pinned versions on package.json, and the CI check that will fail if you are committing a tampered bundle, I think we are well covered when the question comes.

In any case, I've added this sub bullet to clarify a bit:

```